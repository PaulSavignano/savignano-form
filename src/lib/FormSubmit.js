import React, { useMemo } from 'react'
import PropTypes from 'prop-types'

import useFormSubmit from './useFormSubmit'

function FormSubmit(props) {
  const { isErrors, isSubmitSuccess, isSubmitting, onSubmit, submitError } = useFormSubmit()
  const {
    children,
    component: Comp,
    disabled,
    submitStateComponent: SubmitStateComponent,
    onClick,
    onPress,
    ...rest
  } = props
  return useMemo(() => {
    return (
      <Comp
        {...rest}
        {...onClick && { onClick: onSubmit }}
        {...onPress && { onPress: onSubmit }}
        disabled={disabled || isSubmitting || isErrors}
      >
        {SubmitStateComponent ? (
          <SubmitStateComponent
            isSubmitError={Boolean(submitError)}
            isSubmitSuccess={isSubmitSuccess}
            isSubmitting={isSubmitting}
          >
            {children}
          </SubmitStateComponent>
        ) : children}
      </Comp>
    )
  }, [
    SubmitStateComponent,
    children,
    disabled,
    isErrors,
    isSubmitSuccess,
    isSubmitting,
    onClick,
    onPress,
    onSubmit,
    rest,
    submitError,
  ])
}

FormSubmit.defaultProps = {
  component: 'button',
  disabled: false,
  onClick: undefined,
  onPress: undefined,
  submitStateComponent: undefined,
  type: 'submit',
}

FormSubmit.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.string]).isRequired,
  component: PropTypes.oneOfType([PropTypes.node, PropTypes.string, PropTypes.object, PropTypes.func]),
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  onPress: PropTypes.func,
  submitStateComponent: PropTypes.elementType,
  type: PropTypes.string,
}

export default FormSubmit
