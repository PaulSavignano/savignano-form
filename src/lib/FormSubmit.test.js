import React from 'react'
import { shallow } from 'enzyme'

import FormSubmit, { getSubmitState } from './FormSubmit'

const testCtx = {
  isErrors: false,
  isSubmitSuccess: false,
  isSubmitting: false,
  onSubmit: jest.fn(),
  submitError: '',
}

const testProps = {
  children: 'Save',
  component: p => <button type="button" {...p} />,
  submitStateComponent: undefined,
}

describe('getSubmitState', () => {
  it('should return submitting if isSubmitting', () => {
    const submitState = getSubmitState({
      isSubmitting: true
    })
    expect(submitState).toEqual('...submitting')
  })
  it('should return children if isSubmitSuccess', () => {
    const children = 'Submit Form'
    const submitState = getSubmitState({
      children,
      isSubmitSuccess: true
    })
    expect(submitState).toEqual(children)
  })
  it('should return error if isSubmitError', () => {
    const submitState = getSubmitState({
      isSubmitError: true
    })
    expect(submitState).toEqual('Error')
  })
  it('should return children as default', () => {
    const children = 'Submit Form'
    const submitState = getSubmitState({
      children,
      isSubmitError: false,
      isSubmitSuccess: false,
      isSubmitting: false,
    })
    expect(submitState).toEqual(children)
  })
})

describe('FormSubmit', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<FormSubmit {...testProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should create onClick prop with a value of onSubmit when type is button', () => {
    const spy = jest.fn()
    const wrapper = shallow(<FormSubmit {...testProps} type="button" />)
    const contextWrapper = shallow(wrapper.prop('children')({ ...testCtx, onSubmit: spy }))
    const button = contextWrapper.find('button')
    expect(button.props().onClick).toEqual(spy)
  })

  it('should not render with submitStateComponent if NOT defined', () => {
    const wrapper = shallow(<FormSubmit {...testProps} />)
    const contextWrapper = shallow(wrapper.prop('children')({ ...testCtx }))
    expect(contextWrapper.find('submitStateComponent').exists()).toBe(false)
  })

  it('should render submitStateComponent if defined', () => {
    const submitStateComponent = props => <div {...props} />
    const wrapper = shallow(<FormSubmit {...testProps} submitStateComponent={submitStateComponent} />)
    const contextWrapper = shallow(wrapper.prop('children')({ ...testCtx }))
    expect(contextWrapper.find('button').props().disabled).toEqual(false)
    expect(contextWrapper.find('button').props().type).toEqual('submit')
    expect(contextWrapper.find('submitStateComponent').exists()).toBe(true)
  })


})
