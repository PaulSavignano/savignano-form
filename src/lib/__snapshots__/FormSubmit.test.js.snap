// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSubmit should match snapshot 1`] = `ShallowWrapper {}`;
